plugins/experimental/brotli_transform/brotli_transform.cc

```diff
@@ -0,0 +1,248 @@
+    setOutputComplete();
+  }
+
+  virtual ~BrotliTransformationPlugin() {}
+private:
+  void
+  checkContentEncoding()
+  {
+    Headers &hdr           = transaction_.getServerResponse().getHeaders();
+    string contentEncoding = hdr.values("Content-Encoding");
+    if (contentEncoding.empty()) {
+      osContentEncoding_ = PLAINTEXT;
+    } else {
+      if (contentEncoding.find("gzip") != string::npos) {
+        osContentEncoding_ = GZIP;
+        gzipState_         = new GzipInflateTransformationState(RESPONSE_TRANSFORMATION);
+      } else {
+        osContentEncoding_ = OTHER;
+      }
+    }
+  }
+
+  string
+  gzipInflateData(const string &data)
+  {
+    string inflateStr = "";
+    if (data.size() == 0) {
+      return "";
+    }
+
+    if (!gzipState_->z_stream_initialized_) {
+      TS_ERROR(TAG, "Unable to inflate output because the z_stream was not initialized.");
+      return "";
+    }
+
+    int err                = Z_OK;
+    int iteration          = 0;
+    int inflate_block_size = INFLATE_SCALE_FACTOR * data.size();
+    vector<char> buffer(inflate_block_size);
+
+    // Setup the compressed input
+    gzipState_->z_stream_.next_in  = reinterpret_cast<unsigned char *>(const_cast<char *>(data.c_str()));
+    gzipState_->z_stream_.avail_in = data.size();
+
+    // Loop while we have more data to inflate
+    while (gzipState_->z_stream_.avail_in > 0 && err != Z_STREAM_END) {
+      TS_DEBUG(TAG, "Iteration %d: Gzip has %d bytes to inflate", ++iteration, gzipState_->z_stream_.avail_in);
+
+      // Setup where the decompressed output will go.
+      gzipState_->z_stream_.next_out  = reinterpret_cast<unsigned char *>(&buffer[0]);
+      gzipState_->z_stream_.avail_out = inflate_block_size;
+
+      /* Uncompress */
+      err = inflate(&gzipState_->z_stream_, Z_SYNC_FLUSH);
+
+      if (err != Z_OK && err != Z_STREAM_END) {
+        TS_ERROR(TAG, "Iteration %d: Inflate failed with error '%d'", iteration, err);
+        gzipState_->z_stream_.next_out = NULL;
+        return "";
+      }
+
+      TS_DEBUG(TAG, "Iteration %d: Gzip inflated a total of %d bytes, producingOutput...", iteration,
+               (inflate_block_size - gzipState_->z_stream_.avail_out));
+      inflateStr.append(string(&buffer[0], (inflate_block_size - gzipState_->z_stream_.avail_out)));
+      gzipState_->bytes_produced_ += (inflate_block_size - gzipState_->z_stream_.avail_out);
+    }
+    gzipState_->z_stream_.next_out = NULL;
+    return inflateStr;
+  }
+
+  Transaction &transaction_;
+  string buffer_;
+  string outputData_;
+  GzipInflateTransformationState *gzipState_;
+  enum ContentEncoding { GZIP, PLAINTEXT, OTHER };
+  ContentEncoding osContentEncoding_;
+};
+
+class GlobalHookPlugin : public GlobalPlugin
+{
+public:
+  GlobalHookPlugin() { registerHook(HOOK_READ_RESPONSE_HEADERS); }
+  void
+  handleReadResponseHeaders(Transaction &transaction)
+  {
+    if (isBrotliSupported(transaction)) {
+      TS_DEBUG(TAG, "Brotli is supported.");
```

Comment:
Here you could add:

if (responseGzipped) { 
  transaction.addPlugin( new GzipInflafeTransformation(...);
}